### PR TITLE
Examples/minimal rl py

### PR DESCRIPTION
**Pull request by Scitator**
## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/drive/1JCGTVvWlrIsLXMPRRRSWiAstSLic4nbA) with `latest` and `minimal` requirements?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### FAQ
Please review the FAQ before submitting an issue:
- [ ] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [ ] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [ ] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [ ] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [ ] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

### CODE REVIEW

File: examples/reinforcement_learning/ddpg.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+from torch.utils.data import DataLoader
+from torch.utils.data.dataset import IterableDataset
+
+from catalyst import dl, utils
+
+# Off-policy common
+
+Transition = namedtuple(
+    "Transition", field_names=["state", "action", "reward", "done", "next_state"]
+)
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:18]: I'd appreciate it if you could use [dataclasses](https://docs.python.org/3/library/dataclasses.html) here

File: examples/reinforcement_learning/ddpg.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+    def sample(self, size: int) -> Sequence[np.array]:
+        indices = np.random.choice(len(self.buffer), size, replace=size > len(self.buffer))
+        states, actions, rewards, dones, next_states = zip(*[self.buffer[idx] for idx in indices])
+        states, actions, rewards, dones, next_states = (
+            np.array(states, dtype=np.float32),
+            np.array(actions, dtype=np.int64),
+            np.array(rewards, dtype=np.float32),
+            np.array(dones, dtype=np.bool),
+            np.array(next_states, dtype=np.float32),
+        )
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py]: In my opinion it will be a little bit easier to read this
```python
states = np.array(states, dtype=np.float32)
# ...
next_states = np.array(newxt_states, dtype=np.float32)
```

File: examples/reinforcement_learning/ddpg.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+    def __iter__(self) -> Iterator[Sequence[np.array]]:
+        states, actions, rewards, dones, next_states = self.buffer.sample(self.epoch_size)
+        for i in range(len(dones)):
+            yield states[i], actions[i], rewards[i], dones[i], next_states[i]
+
+    def __len__(self) -> int:
+        return self.epoch_size
+
+
+def soft_update(target: nn.Module, source: nn.Module, tau: float):
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py]:
```suggestion
def soft_update(target: nn.Module, source: nn.Module, tau: float) -> None:
```

File: examples/reinforcement_learning/ddpg.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    """Updates the target data with smoothing by ``tau``"""
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py]: Could you please add info that `target` tensor will be modified?

File: examples/reinforcement_learning/ddpg.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+        action = 2 * (action - low_bound) / (upper_bound - low_bound) - 1
+        action = np.clip(action, low_bound, upper_bound)
+
+        return action
+
+
+def get_action(
+    env, network: nn.Module, state: np.array, sigma: Optional[float] = None
+) -> np.array:
+    state = torch.tensor(state, dtype=torch.float32).unsqueeze(0)
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:90]: Could you please add info about the expected shapes of the inputs?

File: examples/reinforcement_learning/ddpg.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    action = network(state).detach().cpu().numpy()[0]
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:91]: `.detach().cpu().numpy()` can be replaced with `catalyst.utils.detach()`

File: examples/reinforcement_learning/ddpg.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+
+    for t in range(env.spec.max_episode_steps):
+        action = get_action(env, network, state=state, sigma=sigma)
+        next_state, reward, done, _ = env.step(action)
+
+        if replay_buffer is not None:
+            transition = Transition(state, action, reward, done, next_state)
+            replay_buffer.append(transition)
+
+        total_reward += reward
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:114]: Do we use a discount factor only during training and that's why we compute total reward without it?

File: examples/reinforcement_learning/ddpg.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        sessions_steps += t
+    return sessions_reward, sessions_steps
+
+
+def get_network_actor(env):
+    inner_fn = utils.get_optimal_inner_init(nn.ReLU)
+    outer_fn = utils.outer_init
+
+    network = torch.nn.Sequential(
+        nn.Linear(env.observation_space.shape[0], 400), nn.ReLU(), nn.Linear(400, 300), nn.ReLU(),
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:144]: Maybe we should pass nn dimensions ( `400` and `300`) as a parameters?

File: examples/reinforcement_learning/ddpg.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+
+    return torch.nn.Sequential(network, head)
+
+
+def get_network_critic(env):
+    inner_fn = utils.get_optimal_inner_init(nn.LeakyReLU)
+    outer_fn = utils.outer_init
+
+    network = torch.nn.Sequential(
+        nn.Linear(env.observation_space.shape[0] + 1, 400),
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:159]: Why `+1` is used here?

File: examples/reinforcement_learning/ddpg.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+    return torch.nn.Sequential(network, head)
+
+
+# Catalyst
+
+
+class GameCallback(dl.Callback):
+    def __init__(
+        self,
+        *,
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:178]: What are the main reasons to force users to use key-value?

File: examples/reinforcement_learning/ddpg.py [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+    def on_stage_start(self, runner: dl.IRunner):
+        self.actor = runner.model[self.actor_key]
+
+        self.actor.eval()
+        generate_sessions(
+            env=self.env,
+            network=self.actor,
+            sigma=self.sigma,
+            replay_buffer=self.replay_buffer,
+            num_sessions=1000,
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py]: Why `1000` is hardcoded?

File: examples/reinforcement_learning/ddpg.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+        )
+        self.actor.train()
+
+    def on_epoch_start(self, runner: dl.IRunner):
+        self.session_counter = 0
+        self.session_steps = 0
+
+    def on_batch_end(self, runner: dl.IRunner):
+        if runner.global_batch_step % self.session_period == 0:
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py:216]:
```suggestion
        if self.session_period is not None and runner.global_batch_step % self.session_period == 0:
```

File: examples/reinforcement_learning/ddpg.py [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+            self.session_counter += 1
+            self.session_steps += session_steps
+
+            runner.batch_metrics.update({"s_reward": session_reward})
+            runner.batch_metrics.update({"s_steps": session_steps})
+
+            self.actor.train()
+
+    def on_epoch_end(self, runner: dl.IRunner):
+        num_sessions = 100
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py]: Why `100` is hardcoded?

File: examples/reinforcement_learning/ddpg.py [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+            actor_optimizer.step()
+
+            critic.zero_grad()
+            critic_optimizer.zero_grad()
+            value_loss.backward()
+            critic_optimizer.step()
+
+            if self.global_batch_step % self.tau_period == 0:
+                soft_update(target_actor, actor, self.tau)
+                soft_update(target_critic, critic, self.tau)
```
Comment (bagxi) [inline examples/reinforcement_learning/ddpg.py]: Is it possible to replace this with an optimizer callback?

File: examples/reinforcement_learning/ddpg.py (hunk 1 above)
Comment (Scitator) [reply examples/reinforcement_learning/ddpg.py:18]: there are now dataclasses under python 3.6 :)

File: examples/reinforcement_learning/ddpg.py (hunk 7 above)
Comment (Scitator) [reply examples/reinforcement_learning/ddpg.py:114]: yes, that's correct

File: examples/reinforcement_learning/ddpg.py (hunk 14 above)
Comment (Scitator) [reply examples/reinforcement_learning/ddpg.py]: yes, but for education purposes it's better to have everything in one place

File: examples/reinforcement_learning/ddpg.py (hunk 12 above)
Comment (Scitator) [reply examples/reinforcement_learning/ddpg.py:216]: session_period shouldn't be None in any case 🤔

File: examples/reinforcement_learning/ddpg.py (hunk 8 above)
Comment (Scitator) [reply examples/reinforcement_learning/ddpg.py:144]: it's the default architecture, so let's keep it simple
